### PR TITLE
Add `Move`

### PR DESCRIPTION
**Pull request by swcraig**
The first commit is changing the method signature for `Pawn#possible_moves`. From the commit message:

In order to iterate all possible moves, pieces (with the exception of
`King`, `Knight` and `Pawn`) must know the board length in order to
know when to stop iteration.  Since it doesn't really make sense for
pieces to know the board length the constant lives in
`Move`.  `Move` will be calling `Piece#possible_moves` with three
arguments, and so `Pawn` needs to disregard the third `side_length`
argument.

Second commit actually adds the `Move` class:

This class is responsible for validating moves. In general,
`Move#valid?` checks that:
  - the source and destination cells are both on the board
  - there is a piece at the source cell
  - the piece at the destination cell belongs to the opposing player
  - if the moving piece is a pawn, it is making a forward move
  - that the moving piece can get to the destination, and
  - there is an open path between the source and destination

Instantiation is `Move#new(board, source_cell, destination_cell)`

Also, exclude the `CyclomaticComplexity` cop from running on `Move`.
`Move#valid?` is upsetting it by one mark, and I don't see an obvious
fix.

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -5,12 +5,18 @@
...
 
   def initialize(player_one, player_two)
     @player_one = player_one
     @player_two = player_two
     @cells = InitializeBoard.setup_pieces(player_one, player_two)
+    @sides = [player_one, player_two]
+  end
+
+  def piece_at(cell)
+    cells[cell[1]][cell[0]]
```
Comment (brendandeere) [inline lib/board.rb]: Im still not thrilled with the idea of passing around arrays of two numbers.... it just feels messy to me

File: lib/move.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Represents a possible move on the @board
+# This class is responsible for validating moves
+class Move
+  SIDE_LENGTH = 7
+
+  def initialize(board, src, dst)
+    @board = board
+    @piece = board.piece_at(src)
+    @src = src
+    @dst = dst
```
Comment (jacquesporveau) [inline lib/move.rb]: I don't know what `@dst` means. I wonder if we can give this variable a more descriptive name.

File: lib/move.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    path_between(@src[0], @src[1]).all?(&:nil?)
+  end
+
+  def path_between(x, y)
+    Array.new(@distance) { @board.piece_at([x += @dx, y += @dy]) }
+  end
+
+  def moves_on_board(moves)
+    moves.reject do |c|
+      c.flatten.any? { |p| p.negative? || p > SIDE_LENGTH }
```
Comment (jacquesporveau) [inline lib/move.rb]: I am not huge on these single letter variable names. What is `c`?

File: spec/move_spec.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      subject { Move.new(board, [1, 1], [1, 3]).valid? }
+
+      it 'returns true' do
+        expect(subject).to be_truthy
+      end
+    end
+
+    context 'a invalid move' do
+      subject { Move.new(board, [1, 1], [1, -1]).valid? }
+      it 'returns false' do
```
Comment (jacquesporveau) [inline spec/move_spec.rb]: There is some inconsistency in your whitespace surrounding your `it` blocks. Personally I like a space above `it` blocks but feel free to choose your own style. I see a line of whitespace above line 23 but not here.

**Comment by jacquesporveau** [comment]
@swcraig Can you please provide a reason you are closing your PRs before you close them?

**Comment by swcraig** [comment]
@jacquesporveau  This PR was based on the approach of getting Pieces to explicitly send `Move` the cells that they could move to from a specific source cell.  We are going to change the approach to one in which a `Piece` will send `Move` the valid movement vectors that they can travel along instead.  This changes much of the code, and makes it necessary for me to make some changes to other classes that this PR relies on.  I will re-open this PR with the updated approach once I get to "re-adding" `Move`.  There are some things I need to re-do before that.

**Comment by jacquesporveau** [comment]
Sounds good! It's nice to leave a paper trail for these sorts of things. Cheers!

**Pull request closed without merging**